// @uiw/react-md-editor/lib/commands/title2.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title2 = void 0;
var React = _interopRequireWildcard(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title2 = {
  name: 'title2',
  keyCommand: 'title2',
  shortcuts: 'ctrlcmd+2',
  value: 'title2',
  buttonProps: {
    'aria-label': 'Insert title2 (ctrl + 2)',
    title: 'Insert title2 (ctrl + 2)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 16,
      textAlign: 'left'
    },
    children: "Title 2"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('## ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('## ', state.selection.start, api.textArea);
    }
  }
};
exports.title2 = title2;
