// @uiw/react-md-editor/lib/commands/title1.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title1 = void 0;
var _react = _interopRequireDefault(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title1 = {
  name: 'title1',
  keyCommand: 'title1',
  shortcuts: 'ctrlcmd+1',
  value: 'title1',
  buttonProps: {
    'aria-label': 'Insert title1 (ctrl + 1)',
    title: 'Insert title1 (ctrl + 1)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 18,
      textAlign: 'left'
    },
    children: "Title 1"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('# ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('# ', state.selection.start, api.textArea);
    }
  }
};
exports.title1 = title1;
