// @uiw/react-md-editor/lib/commands/title4.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title4 = void 0;
var React = _interopRequireWildcard(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title4 = {
  name: 'title4',
  keyCommand: 'title4',
  shortcuts: 'ctrlcmd+4',
  value: 'title4',
  buttonProps: {
    'aria-label': 'Insert title4 (ctrl + 4)',
    title: 'Insert title4 (ctrl + 4)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 14,
      textAlign: 'left'
    },
    children: "Title 4"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('#### ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('#### ', state.selection.start, api.textArea);
    }
  }
};
exports.title4 = title4;
