// @uiw/react-md-editor/lib/commands/title6.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title6 = void 0;
var React = _interopRequireWildcard(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title6 = {
  name: 'title6',
  keyCommand: 'title6',
  shortcuts: 'ctrlcmd+6',
  value: 'title6',
  buttonProps: {
    'aria-label': 'Insert title6 (ctrl + 6)',
    title: 'Insert title6 (ctrl + 6)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 12,
      textAlign: 'left'
    },
    children: "Title 6"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('###### ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('###### ', state.selection.start, api.textArea);
    }
  }
};
exports.title6 = title6;
