// @uiw/react-md-editor/lib/commands/title3.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title3 = void 0;
var React = _interopRequireWildcard(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title3 = {
  name: 'title3',
  keyCommand: 'title3',
  shortcuts: 'ctrlcmd+3',
  value: 'title3',
  buttonProps: {
    'aria-label': 'Insert title3 (ctrl + 3)',
    title: 'Insert title3 (ctrl + 3)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 15,
      textAlign: 'left'
    },
    children: "Title 3"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('### ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('### ', state.selection.start, api.textArea);
    }
  }
};
exports.title3 = title3;
