// @uiw/react-md-editor/lib/commands/title5.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.title5 = void 0;
var React = _interopRequireWildcard(require("react"));
var _InsertTextAtPosition = require("../utils/InsertTextAtPosition");
var _jsxRuntime = require("react/jsx-runtime");
var title5 = {
  name: 'title5',
  keyCommand: 'title5',
  shortcuts: 'ctrlcmd+5',
  value: 'title5',
  buttonProps: {
    'aria-label': 'Insert title5 (ctrl + 5)',
    title: 'Insert title5 (ctrl + 5)'
  },
  icon: /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
    style: {
      fontSize: 12,
      textAlign: 'left'
    },
    children: "Title 5"
  }),
  execute: function execute(state, api) {
    if (state.selection.start === 0 || /\n$/.test(state.text)) {
      api.replaceSelection('##### ');
    } else {
      (0, _InsertTextAtPosition.insertAtLineStart)('##### ', state.selection.start, api.textArea);
    }
  }
};
exports.title5 = title5;
